# <center> **Car Price Analysis - Marketing**

## Objectives

In this Notebook Explanotory Data Analysis (EDA) and Vizualisation Analysis are performed on the "Car Price Prediction" data set [Kaggle](https://www.kaggle.com/datasets/hellbuoy/car-price-prediction). 



---

For this project different Python librarires are used for analysis and vizualisation. Libraries are imported prior furtherwork on the project.

In [1]:
import pandas as pd                 #import Pandas for data manipulation
import numpy as np                  #import Numpy for numerical operations
import matplotlib.pyplot as plt     #import Matplotlib for data visualization
import seaborn as sns               #import Seaborn for statistical data visualization
import plotly.express as px         #import Plotly Express for interactive visualizations

Style and plot size are set

In [2]:
sns.set(style="whitegrid")                  # Set Seaborn style for plots
plt.rcParams["figure.figsize"] = (10,6)     # Set default figure size for Matplotlib plots

### ***1. Explanatory Data Analysis***

##### In this section EDA, including data load and cleaning, is performed.

As a first step, "CarPrice_Assignment.csv" data set is loaded into DataFrame

In [3]:
df = pd.read_csv('../Data/CarPrice_Assignment.csv')  # Load the car price dataset
df.head()                                            # Display the first few rows of the dataset

car_ID  symboling                   CarName fueltype aspiration doornumber  \
0       1          3        alfa-romero giulia      gas        std        two   
1       2          3       alfa-romero stelvio      gas        std        two   
2       3          1  alfa-romero Quadrifoglio      gas        std        two   
3       4          2               audi 100 ls      gas        std       four   
4       5          2                audi 100ls      gas        std       four   

       carbody drivewheel enginelocation  wheelbase  ...  enginesize  \
0  convertible        rwd          front       88.6  ...         130   
1  convertible        rwd          front       88.6  ...         130   
2    hatchback        rwd          front       94.5  ...         152   
3        sedan        fwd          front       99.8  ...         109   
4        sedan        4wd          front       99.4  ...         136   

   fuelsystem  boreratio  stroke compressionratio horsepower  peakrpm citympg  \
0        mpfi       3.47    2.68              9.0        111     5000      21   
1        mpfi       3.47    2.68              9.0        111     5000      21   
2        mpfi       2.68    3.47              9.0        154     5000      19   
3        mpfi       3.19    3.40             10.0        102     5500      24   
4        mpfi       3.19    3.40              8.0        115     5500      18   

   highwaympg    price  
0          27  13495.0  
1          27  16500.0  
2          26  16500.0  
3          30  13950.0  
4          22  17450.0  

[5 rows x 26 columns]

##### ***1.1 Initial data exploration***

*In the following subsection initial data set inspection is performed*

Here the shape and Info of DataFrame are shown

In [4]:
print(df.shape)                     # Print the shape of the DataFrame           
print(df.info())                    # Print concise summary of the DataFrame            
print(df.dtypes)                    # Print data types of each column

(205, 26)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   car_ID            205 non-null    int64  
 1   symboling         205 non-null    int64  
 2   CarName           205 non-null    object 
 3   fueltype          205 non-null    object 
 4   aspiration        205 non-null    object 
 5   doornumber        205 non-null    object 
 6   carbody           205 non-null    object 
 7   drivewheel        205 non-null    object 
 8   enginelocation    205 non-null    object 
 9   wheelbase         205 non-null    float64
 10  carlength         205 non-null    float64
 11  carwidth          205 non-null    float64
 12  carheight         205 non-null    float64
 13  curbweight        205 non-null    int64  
 14  enginetype        205 non-null    object 
 15  cylindernumber    205 non-null    object 
 16  enginesize        205 non-null    

As it can been, DataFrame consists of 205 etries and 26 columns with following types:
* float64(8)
* int64(8)
* object(10)

In the next steps DataFrame is checked for any incosistencies(dublicates, missing value and etc.)

In [5]:
df.isnull().sum()                 # Check for missing values in each column


car_ID              0
symboling           0
CarName             0
fueltype            0
aspiration          0
doornumber          0
carbody             0
drivewheel          0
enginelocation      0
wheelbase           0
carlength           0
carwidth            0
carheight           0
curbweight          0
enginetype          0
cylindernumber      0
enginesize          0
fuelsystem          0
boreratio           0
stroke              0
compressionratio    0
horsepower          0
peakrpm             0
citympg             0
highwaympg          0
price               0
dtype: int64

as we can see above the DataFrame has no missing values. Let's check for duplicates:

In [6]:
df.duplicated() # Check for duplicate rows in the DataFrame


0      False
1      False
2      False
3      False
4      False
       ...  
200    False
201    False
202    False
203    False
204    False
Length: 205, dtype: bool

And there are no duplicates. Initial data inspection shows that DataFrame has no missing values and duplicates, which simplifies further work with Data.

In following cell a descriptive statistics of numeric columns is shown

In [7]:
df.describe() # Generate descriptive statistics of numerical columns

car_ID   symboling   wheelbase   carlength    carwidth   carheight  \
count  205.000000  205.000000  205.000000  205.000000  205.000000  205.000000   
mean   103.000000    0.834146   98.756585  174.049268   65.907805   53.724878   
std     59.322565    1.245307    6.021776   12.337289    2.145204    2.443522   
min      1.000000   -2.000000   86.600000  141.100000   60.300000   47.800000   
25%     52.000000    0.000000   94.500000  166.300000   64.100000   52.000000   
50%    103.000000    1.000000   97.000000  173.200000   65.500000   54.100000   
75%    154.000000    2.000000  102.400000  183.100000   66.900000   55.500000   
max    205.000000    3.000000  120.900000  208.100000   72.300000   59.800000   

        curbweight  enginesize   boreratio      stroke  compressionratio  \
count   205.000000  205.000000  205.000000  205.000000        205.000000   
mean   2555.565854  126.907317    3.329756    3.255415         10.142537   
std     520.680204   41.642693    0.270844    0.313597          3.972040   
min    1488.000000   61.000000    2.540000    2.070000          7.000000   
25%    2145.000000   97.000000    3.150000    3.110000          8.600000   
50%    2414.000000  120.000000    3.310000    3.290000          9.000000   
75%    2935.000000  141.000000    3.580000    3.410000          9.400000   
max    4066.000000  326.000000    3.940000    4.170000         23.000000   

       horsepower      peakrpm     citympg  highwaympg         price  
count  205.000000   205.000000  205.000000  205.000000    205.000000  
mean   104.117073  5125.121951   25.219512   30.751220  13276.710571  
std     39.544167   476.985643    6.542142    6.886443   7988.852332  
min     48.000000  4150.000000   13.000000   16.000000   5118.000000  
25%     70.000000  4800.000000   19.000000   25.000000   7788.000000  
50%     95.000000  5200.000000   24.000000   30.000000  10295.000000  
75%    116.000000  5500.000000   30.000000   34.000000  16503.000000  
max    288.000000  6600.000000   49.000000   54.000000  45400.000000

Summary od descriptive statistics:
* The dataset spans a wide spectrum of cars, not just one segment.
* Horsepower, engine size, and curb weight strongly influence price.
* Data is clean, numeric-heavy, and ready for transformation.

---

##### ***1.2 Data cleaning and preparation***

##### *In this subsection basic data cleaning and preparation for visualizations are performed*

Copy original DataFrame

In [8]:
df_cleaned = df.copy()          # Create a copy of the original DataFrame for cleaning

Fetching all columns names

In [9]:
df_cleaned.columns          # Display the column names of the DataFrame

Index(['car_ID', 'symboling', 'CarName', 'fueltype', 'aspiration',
       'doornumber', 'carbody', 'drivewheel', 'enginelocation', 'wheelbase',
       'carlength', 'carwidth', 'carheight', 'curbweight', 'enginetype',
       'cylindernumber', 'enginesize', 'fuelsystem', 'boreratio', 'stroke',
       'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg',
       'price'],
      dtype='object')

For further analysis "car_ID" column can be dropped

In [10]:
df_cleaned.drop("car_ID",axis=1,inplace=True)
df_cleaned.head()

symboling                   CarName fueltype aspiration doornumber  \
0          3        alfa-romero giulia      gas        std        two   
1          3       alfa-romero stelvio      gas        std        two   
2          1  alfa-romero Quadrifoglio      gas        std        two   
3          2               audi 100 ls      gas        std       four   
4          2                audi 100ls      gas        std       four   

       carbody drivewheel enginelocation  wheelbase  carlength  ...  \
0  convertible        rwd          front       88.6      168.8  ...   
1  convertible        rwd          front       88.6      168.8  ...   
2    hatchback        rwd          front       94.5      171.2  ...   
3        sedan        fwd          front       99.8      176.6  ...   
4        sedan        4wd          front       99.4      176.6  ...   

   enginesize  fuelsystem  boreratio stroke compressionratio  horsepower  \
0         130        mpfi       3.47   2.68              9.0         111   
1         130        mpfi       3.47   2.68              9.0         111   
2         152        mpfi       2.68   3.47              9.0         154   
3         109        mpfi       3.19   3.40             10.0         102   
4         136        mpfi       3.19   3.40              8.0         115   

  peakrpm  citympg  highwaympg    price  
0    5000       21          27  13495.0  
1    5000       21          27  16500.0  
2    5000       19          26  16500.0  
3    5500       24          30  13950.0  
4    5500       18          22  17450.0  

[5 rows x 25 columns]

Let's take a close look on "CarName" column.The column "brand" is extracted, car brands are uniformly renamed and old "CarName" column is dropped.

In [11]:
        
df_cleaned['brand'] = df_cleaned['CarName'].apply(lambda x: str(x).split(' ')[0].lower())   # Extract car brand from "CarName" column                                        
df_cleaned = df_cleaned.drop(['CarName'], axis=1)                                           # Drop the original "CarName" column
df_cleaned.head()                                                                                   

symboling fueltype aspiration doornumber      carbody drivewheel  \
0          3      gas        std        two  convertible        rwd   
1          3      gas        std        two  convertible        rwd   
2          1      gas        std        two    hatchback        rwd   
3          2      gas        std       four        sedan        fwd   
4          2      gas        std       four        sedan        4wd   

  enginelocation  wheelbase  carlength  carwidth  ...  fuelsystem  boreratio  \
0          front       88.6      168.8      64.1  ...        mpfi       3.47   
1          front       88.6      168.8      64.1  ...        mpfi       3.47   
2          front       94.5      171.2      65.5  ...        mpfi       2.68   
3          front       99.8      176.6      66.2  ...        mpfi       3.19   
4          front       99.4      176.6      66.4  ...        mpfi       3.19   

  stroke compressionratio  horsepower peakrpm  citympg  highwaympg    price  \
0   2.68              9.0         111    5000       21          27  13495.0   
1   2.68              9.0         111    5000       21          27  16500.0   
2   3.47              9.0         154    5000       19          26  16500.0   
3   3.40             10.0         102    5500       24          30  13950.0   
4   3.40              8.0         115    5500       18          22  17450.0   

         brand  
0  alfa-romero  
1  alfa-romero  
2  alfa-romero  
3         audi  
4         audi  

[5 rows x 25 columns]

Earlier, "brand" column have been checked for duplicates, but let's take a closer look on unique brand names.

In [12]:
df_cleaned ['brand'].unique()  # Check unique values in the 'brand' column

array(['alfa-romero', 'audi', 'bmw', 'chevrolet', 'dodge', 'honda',
       'isuzu', 'jaguar', 'maxda', 'mazda', 'buick', 'mercury',
       'mitsubishi', 'nissan', 'peugeot', 'plymouth', 'porsche',
       'porcshce', 'renault', 'saab', 'subaru', 'toyota', 'toyouta',
       'vokswagen', 'volkswagen', 'vw', 'volvo'], dtype=object)

"Brand" column contains the same names, but with typos or different way of spelling, e.g. 'maxda'and 'mazda', 'toyota' and 'toyouta' and etc. It is necessary to replace duplicates.

In [13]:
doubles = {'maxda':'mazda','porcshce':'porsche','toyouta':'toyota',
         'vokswagen':'volkswagen','vw':'volkswagen'}    # Dictionary to map duplicate brand names to correct ones                                                                    
df_cleaned['brand'] = df_cleaned['brand'].replace(doubles)              # Replace duplicate brand names with correct ones   

Changes verification

In [14]:
df_cleaned['brand'].unique()  # Check unique values in the 'brand' column again to confirm changes

array(['alfa-romero', 'audi', 'bmw', 'chevrolet', 'dodge', 'honda',
       'isuzu', 'jaguar', 'mazda', 'buick', 'mercury', 'mitsubishi',
       'nissan', 'peugeot', 'plymouth', 'porsche', 'renault', 'saab',
       'subaru', 'toyota', 'volkswagen', 'volvo'], dtype=object)

Now we check "doornumber" column for unique names.

In [15]:
df_cleaned ['doornumber'].unique()      # Check unique values in the 'doornumber' column    

array(['two', 'four'], dtype=object)

Replace text features with numbers and set column type as integer

In [16]:
df_cleaned['doornumber'] = df_cleaned['doornumber'].str.strip().str.lower()  # Clean up whitespace and convert to lowercase

door_map = {
    'two': 2,
    'four': 4 
    }         # Create a mapping dictionary


df_cleaned['doornumber'] = df_cleaned['doornumber'].replace(door_map).astype(int)# Replace and cast to int


C:\Users\julia\AppData\Local\Temp\ipykernel_35780\42335584.py:9: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_cleaned['doornumber'] = df_cleaned['doornumber'].replace(door_map).astype(int)# Replace and cast to int


And check for resultes

In [17]:
print(df_cleaned['doornumber'].unique())
print(df_cleaned['doornumber'].dtype)

[2 4]
int64


Also number of cylinders were chaged to numbers, and dtype to integer

In [18]:
print(df_cleaned['cylindernumber'].unique()) # Check unique values in the 'cylindernumber' column                                        

['four' 'six' 'five' 'three' 'twelve' 'two' 'eight']


In [19]:
# Clean up whitespace and convert to lowercase
df_cleaned['cylindernumber'] = (
    df_cleaned['cylindernumber']
      .astype(str)
      .str.strip()
      .str.lower()
)
# Create a mapping dictionary
cylinder_map = {
    'two': 2,
    'three': 3,
    'four': 4,
    'five': 5,
    'six': 6,
    'eight': 8,
    'twelve': 12
}
# Replace and cast to int
df_cleaned['cylindernumber'] = df_cleaned['cylindernumber'].replace(cylinder_map).astype(int)

C:\Users\julia\AppData\Local\Temp\ipykernel_35780\3464359280.py:19: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_cleaned['cylindernumber'] = df_cleaned['cylindernumber'].replace(cylinder_map).astype(int)


Verify changes

In [20]:
print(df_cleaned['cylindernumber'].unique())
print(df_cleaned['cylindernumber'].dtype)

[ 4  6  5  3 12  2  8]
int64


The cleaning and data preparation phase is finished. Data set is ready for further visualisation in Python and Tableau.

---

# Push .csv file to Repo

Save cleaned data into dedicated reporsitory on GitHub

In [24]:
df_cleaned.to_csv('../Data/car_price_cleaned_data.csv', index=False)
